dcc-download-import/src/main/java/org/icgc/dcc/download/imports/load/ReleaseFileLoader.java

```diff
@@ -0,0 +1,125 @@
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;                               
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER                              
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN                         
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.icgc.dcc.download.imports.load;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.icgc.dcc.download.imports.core.ArchiveFileType.RELEASE;
+import static org.icgc.dcc.download.imports.util.TarArchiveStreams.getTarInputStream;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import lombok.Cleanup;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.val;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.icgc.dcc.download.imports.io.TarArchiveDocumentReaderFactory;
+import org.icgc.dcc.download.imports.io.TarArchiveEntryCallbackContext;
+import org.icgc.dcc.download.imports.io.TarArchiveEntryCallbackFactory;
+import org.icgc.dcc.release.core.document.DocumentType;
+
+import com.google.common.base.Stopwatch;
+
+@Slf4j
+@RequiredArgsConstructor
+public class ReleaseFileLoader implements FileLoader {
+
+  private static final Pattern TYPE_TAR_NAME_PATTERN = Pattern.compile("^(.*)_(.*)\\.tar\\.gz$");
+
+  @Nullable
+  private final String project;
+  @NonNull
+  private final TarArchiveEntryCallbackFactory callbackFactory;
+  @NonNull
+  private final TarArchiveDocumentReaderFactory readerFactory;
+
+  @Override
+  @SneakyThrows
+  public void loadFile(@NonNull File file) {
+    @Cleanup
+    val tarInput = getTarInputStream(file);
+
+    TarArchiveEntry tarEntry;
+    boolean applySettings = true;
+    while ((tarEntry = tarInput.getNextTarEntry()) != null) { // NOPMD
+      processTypeTarEntry(tarInput, tarEntry, applySettings);
+      applySettings = false;
+    }
+  }
+
+  private void processTypeTarEntry(InputStream inputStream, TarArchiveEntry tarEntry, boolean applySettings)
+      throws IOException {
+    val entryName = tarEntry.getName();
+
+    log.debug("Creating tar document reader for tar entry {}", entryName);
+    val reader = readerFactory.createReader(inputStream);
+    val documentType = resolveDocumentType(entryName);
+    val indexName = resolveIndexName(entryName);
+
+    log.info("Indexing file '{}' into index '{}'", entryName, indexName);
+    // Don't use @Cleanup as the callback will be closed after the log message "Finished indexing file" has been
+    // written.
+    val callbackContext = TarArchiveEntryCallbackContext.builder()
```

Comment:
Maybe `TarArchiveEntryContext` would be a more appropriate scope for the concept.